### PR TITLE
Port Intersection algorithm to C++, cleanups and tests

### PR DESCRIPTION
**Pull request by wonder-sk**
The time has come to start fixing QGIS geometry overlay algorithms :-) This PR is just a starting point - bringing Intersection alg to C++, cleaning its code and adding first auto tests. I would like to port other algs (union, difference, sym. difference) as well, use the same set of input data for tests and fix them along the way.

Q: what do we do with the python algorithm "qgis:intersection" - shall we kill it, possibly breaking models, or shall we keep it as a proxy to "native:intersection", or something else?

Starting a suite of carefully crafted test data for overlay algorithms:
- overlay1 - layers that cover various basic overlay situations
- overlay2 - layers where one input has self-intersecting polygons
- overlay3 - layers where intersections return different geometry types

### CODE REVIEW

**Comment by nirvn** [comment]
Yes please!

File: src/analysis/processing/qgsalgorithmintersection.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+      QgsGeometry tmpGeom( featB.geometry() );
+      if ( !engine->intersects( tmpGeom.constGet() ) )
+        continue;
+
+      QgsGeometry intGeom = geom.intersection( tmpGeom );
+
+      if ( intGeom.isNull() )
+      {
+        // TODO: not sure if this ever happens - if it does, that means GEOS failed badly and we should try to provide an error from it
+        throw QgsProcessingException( QObject::tr( "GEOS geoprocessing error: intersection failed." ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmintersection.cpp]: Yes, use intGeon.lastError to get the geos exception text

File: src/analysis/processing/qgsalgorithmintersection.cpp [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+QString QgsIntersectionAlgorithm::group() const
+{
+  return QObject::tr( "Vector overlay" );
+}
+
+QString QgsIntersectionAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectoroverlay" );
+}
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmintersection.cpp]: You'll need to pull the help text from the python alg too

File: src/analysis/processing/qgsalgorithmintersection.cpp [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+#include "qgsgeometryengine.h"
+
+#include "qgsmultipoint.h"
+#include "qgsmultipolygon.h"
+#include "qgsmultilinestring.h"
+
+///@cond PRIVATE
+
+static QgsGeometry geometryCollectionToSingleGeometryType( const QgsGeometry &geom, QgsWkbTypes::GeometryType geomType )
+{
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmintersection.cpp]: How do you feel about moving this to QgsGeometry? I think it's useful outside this alg. Also if it were in QgsGeometry you could likely avoid the clones by doing similar tricks to QgsGeometry::convertToMultipart.

File: src/analysis/processing/qgsalgorithmintersection.cpp [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+  const QgsGeometryCollection *origGeom = qgsgeometry_cast<const QgsGeometryCollection *>( geom.constGet() );
+  if ( !origGeom )
+    return QgsGeometry();
+
+  QgsGeometryCollection *resGeom = nullptr;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmintersection.cpp]: unique_ptr

File: src/analysis/processing/qgsalgorithmintersection.cpp [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+  switch ( geomType )
+  {
+    case QgsWkbTypes::PointGeometry:
+      resGeom = new QgsMultiPoint;
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmintersection.cpp]: qgis::make_unique

File: src/analysis/processing/qgsalgorithmintersection.cpp [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+  if ( !resGeom )
+    return QgsGeometry();
+
+  for ( int i = 0; i < origGeom->numGeometries(); ++i )
+  {
+    const QgsAbstractGeometry *g = origGeom->geometryN( i );
+    if ( QgsWkbTypes::geometryType( g->wkbType() ) == geomType )
+      resGeom->addGeometry( g->clone() );
+  }
+  return QgsGeometry( resGeom );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmintersection.cpp]: std::move( resGeom )

**Review by nyalldawson — APPROVED** [review]
Nice! Always happy to see these algs get some love!

**Comment by nyalldawson** [comment]
> what do we do with the python algorithm "qgis:intersection" - shall we kill it, possibly breaking models, or shall we keep it as a proxy to "native:intersection", or something else

Kill it. Processing silently aliases qgis: algs to native: algs to handle this situation, so no model/script breakage occurs!

FYI there's also https://github.com/nyalldawson/QGIS/wiki/Processing-Algorithm-Bugs-and-Enhancements which collects some thoughts/discussion on issues and potential improvements for these algorithms.

**Comment by m-kuhn** [comment]
Nice, do the dataset also include some null geometries to make sure they are handled properly?

**Comment by wonder-sk** [comment]
@nyalldawson thanks for the review - everything should be addressed now

@m-kuhn no null geometries there yet - that's on my to-do list for "overlay4" test data: invalid geometries + null geometries

**Pull request merged**